### PR TITLE
feat(diagnostics) added profiling for kic 2.0

### LINKED ISSUE #1298: KIC 2.0: Handle profiling flag
1.x uses the `--profiling` flag to enable [pprof](https://golang.org/pkg/net/http/pprof/) endpoints [on the health/metrics/etc. HTTP server](https://github.com/Kong/kubernetes-ingress-controller/blob/v1.2.0/cli/ingress-controller/main.go#L698-L709).

2.x uses the controller-runtime default servers, so it doesn't immediately have a place to put this. We'd need to add a custom server to still expose profiling data. The rest of the implementation is unlikely to change, since we just call pprof's functions.

### PR DESCRIPTION
**Pull request by tharun208**
Signed-off-by: Tharun <rajendrantharun@live.com>

**What this PR does / why we need it**: This PR introduces profiling for kic 2.x

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1298

**Special notes for your reviewer**:
 Created a separate server as when #1278 is implemented it will be easy to add to this change.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: railgun/internal/diagnostics/server.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/kong/kubernetes-ingress-controller/railgun/internal/mgrutils"
+	"github.com/sirupsen/logrus"
+)
+
+const diagnosticsServerPort = 10254
```
Comment (rainest) [inline railgun/internal/diagnostics/server.go:12]: You'll want to move this constant to `railgun/pkg/config/consts.go` and change it to 10256 to not conflict with the other. 1.x served everything on the same port, but we can't do that while using the Kubebuilder default health and metrics system.

File: railgun/internal/diagnostics/server.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			case http.ErrServerClosed:
+				s.logger.Info("shutting down diagnostics server")
+			default:
+				s.logger.Error(err, "could not start a diagnostics server")
+				errChan <- err
+			}
+		}
+	}()
+
+	s.logger.Info("started diagonistics server at port ", diagnosticsServerPort)
```
Comment (rainest) [inline railgun/internal/diagnostics/server.go:46]:
```suggestion
	s.logger.Info("started diagnostics server at port ", diagnosticsServerPort)
```

**Review by rainest — CHANGES_REQUESTED** [review]
Some minor changes. We should also add a basic integration test for this since it's easy to make something similar to the other endpoints. [profiling_integration.diff.txt](https://github.com/Kong/kubernetes-ingress-controller/files/6627364/profiling_integration.diff.txt) works (assuming the other change to move the port constant).

I think we also want to target this to next--need to confirm, since our release process for new 2.x pre-releases is a bit new.

File: railgun/internal/diagnostics/server.go (hunk 2 above)
Comment (tharun208) [reply railgun/internal/diagnostics/server.go:46]: addressed

File: railgun/internal/diagnostics/server.go (hunk 1 above)
Comment (tharun208) [reply railgun/internal/diagnostics/server.go]: addressed

**Comment by rainest** [comment]
Bah, GH UI to change base doesn't allow for a rebase and pulls in other garbage. We do indeed want next for this based on the other 2.x PRs.

@tharun208 the rest all looks good now, so if you can go ahead and rebase onto next this should be good to merge in.

**Comment by tharun208** [comment]
> Bah, GH UI to change base doesn't allow for a rebase and pulls in other garbage. We do indeed want next for this based on the other 2.x PRs.
> 
> @tharun208 the rest all looks good now, so if you can go ahead and rebase onto next this should be good to merge in.

do I need to raise a new pr against the next branch?

**Comment by tharun208** [comment]
> Bah, GH UI to change base doesn't allow for a rebase and pulls in other garbage. We do indeed want next for this based on the other 2.x PRs.
> 
> @tharun208 the rest all looks good now, so if you can go ahead and rebase onto next this should be good to merge in.

done @rainest

**Comment by rainest** [comment]
Latest changes here didn't seem to do what I wanted either, I think from a combination of rebases not inherently changing the target and 72b3542 merging the commits from main back in.

I think you can correct that by re-running the rebase and picking d85f8aadd3007b34ecc025de2b6e3b6239dbabec and 745bfad270c87d2c3d3903eef73a5ed62d99c580, and a953ba7 only, force-pushing, and then changing the base on the PR, which should make those the only three commits in the original PR.

Alternate version is #1416 where I cherry-picked the same into a fresh branch off next, though for that I'll need to find someone else on the team to hit the approve button. Leaving both open until one or the other is merged.

**Comment by rainest** [comment]
Alright, not sure what's happened now. The old main merge is gone, but am now seeing unintended commits from both main _and_ next :shrug:

Will continue on #1416, trying to figure out how to restore order to this version is probably more effort than it's worth.

**Pull request closed without merging**